// @my-react-app$clayui%2Fcore@3.87.0/lib/index.js
"use strict";

Liferay.Loader.define("@my-react-app$clayui/core@3.87.0/lib/index", ['module', 'exports', 'require', '@my-react-app$clayui/button', '@my-react-app$clayui/icon', '@my-react-app$clayui/modal', '@my-react-app$clayui/provider', './typography', './overlay-mask', './tree-view', './vertical-bar', './picker', './collection'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    function _typeof(obj) {
      "@babel/helpers - typeof";
      return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) {
        return typeof obj;
      } : function (obj) {
        return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      }, _typeof(obj);
    }

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    Object.defineProperty(exports, "Button", {
      enumerable: true,
      get: function get() {
        return _button.default;
      }
    });
    Object.defineProperty(exports, "ButtonWithIcon", {
      enumerable: true,
      get: function get() {
        return _button.ClayButtonWithIcon;
      }
    });
    Object.defineProperty(exports, "Heading", {
      enumerable: true,
      get: function get() {
        return _typography.Heading;
      }
    });
    Object.defineProperty(exports, "Icon", {
      enumerable: true,
      get: function get() {
        return _icon.default;
      }
    });
    Object.defineProperty(exports, "Modal", {
      enumerable: true,
      get: function get() {
        return _modal.default;
      }
    });
    Object.defineProperty(exports, "ModalContext", {
      enumerable: true,
      get: function get() {
        return _modal.Context;
      }
    });
    Object.defineProperty(exports, "Option", {
      enumerable: true,
      get: function get() {
        return _picker.Option;
      }
    });
    Object.defineProperty(exports, "OverlayMask", {
      enumerable: true,
      get: function get() {
        return _overlayMask.OverlayMask;
      }
    });
    Object.defineProperty(exports, "Picker", {
      enumerable: true,
      get: function get() {
        return _picker.Picker;
      }
    });
    Object.defineProperty(exports, "Provider", {
      enumerable: true,
      get: function get() {
        return _provider.Provider;
      }
    });
    Object.defineProperty(exports, "Text", {
      enumerable: true,
      get: function get() {
        return _typography.Text;
      }
    });
    Object.defineProperty(exports, "TreeView", {
      enumerable: true,
      get: function get() {
        return _treeView.TreeView;
      }
    });
    Object.defineProperty(exports, "VerticalBar", {
      enumerable: true,
      get: function get() {
        return _verticalBar.VerticalBar;
      }
    });
    exports.__NOT_PUBLIC_COLLECTION = void 0;
    Object.defineProperty(exports, "useModal", {
      enumerable: true,
      get: function get() {
        return _modal.useModal;
      }
    });
    Object.defineProperty(exports, "useProvider", {
      enumerable: true,
      get: function get() {
        return _provider.useProvider;
      }
    });

    var _button = _interopRequireWildcard(require("@my-react-app$clayui/button"));

    var _icon = _interopRequireDefault(require("@my-react-app$clayui/icon"));

    var _modal = _interopRequireWildcard(require("@my-react-app$clayui/modal"));

    var _provider = require("@my-react-app$clayui/provider");

    var _typography = require("./typography");

    var _overlayMask = require("./overlay-mask");

    var _treeView = require("./tree-view");

    var _verticalBar = require("./vertical-bar");

    var _picker = require("./picker");

    var _NOT_PUBLIC_COLLECTION = _interopRequireWildcard(require("./collection"));

    exports.__NOT_PUBLIC_COLLECTION = _NOT_PUBLIC_COLLECTION;

    function _interopRequireDefault(obj) {
      return obj && obj.__esModule ? obj : { default: obj };
    }

    function _getRequireWildcardCache(nodeInterop) {
      if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
      })(nodeInterop);
    }

    function _interopRequireWildcard(obj, nodeInterop) {
      if (!nodeInterop && obj && obj.__esModule) {
        return obj;
      }if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {
        return { default: obj };
      }var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {
        return cache.get(obj);
      }var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
          var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {
            Object.defineProperty(newObj, key, desc);
          } else {
            newObj[key] = obj[key];
          }
        }
      }newObj.default = obj;if (cache) {
        cache.set(obj, newObj);
      }return newObj;
    }
  }
});
