// @my-react-app$clayui%2Fcore@3.87.0/lib/live-announcer/VisuallyHidden.js
"use strict";

Liferay.Loader.define("@my-react-app$clayui/core@3.87.0/lib/live-announcer/VisuallyHidden", ['module', 'exports', 'require', 'my-react-app$react'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.VisuallyHidden = VisuallyHidden;

    var _react = _interopRequireDefault(require("my-react-app$react"));

    function _interopRequireDefault(obj) {
      return obj && obj.__esModule ? obj : { default: obj };
    }

    /**
     * SPDX-FileCopyrightText: © 2022 Liferay, Inc. <https://liferay.com>
     * SPDX-License-Identifier: BSD-3-Clause
     */
    var styles = {
      border: 0,
      clip: 'rect(0 0 0 0)',
      clipPath: 'inset(50%)',
      height: 1,
      margin: '0 -1px -1px 0',
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      whiteSpace: 'nowrap',
      width: 1
    };

    function VisuallyHidden(_ref) {
      var children = _ref.children,
          liveAnnouncer = _ref.liveAnnouncer;
      return (/*#__PURE__*/_react.default.createElement("div", {
          "data-live-announcer": liveAnnouncer,
          style: styles
        }, children)
      );
    }
  }
});
