// my-react-app$react-dnd-html5-backend@11.1.3/lib/NativeDragSources/nativeTypesConfig.js
Liferay.Loader.define('my-react-app$react-dnd-html5-backend@11.1.3/lib/NativeDragSources/nativeTypesConfig', ['module', 'exports', 'require'], function (module, exports, require) {
    var define = undefined;
    var global = window;
    {
        import * as NativeTypes from '../NativeTypes';
        import { getDataFromDataTransfer } from './getDataFromDataTransfer';
        export const nativeTypesConfig = {
            [NativeTypes.FILE]: {
                exposeProperties: {
                    files: dataTransfer => Array.prototype.slice.call(dataTransfer.files),
                    items: dataTransfer => dataTransfer.items
                },
                matchesTypes: ['Files']
            },
            [NativeTypes.URL]: {
                exposeProperties: {
                    urls: (dataTransfer, matchesTypes) => getDataFromDataTransfer(dataTransfer, matchesTypes, '').split('\n')
                },
                matchesTypes: ['Url', 'text/uri-list']
            },
            [NativeTypes.TEXT]: {
                exposeProperties: {
                    text: (dataTransfer, matchesTypes) => getDataFromDataTransfer(dataTransfer, matchesTypes, '')
                },
                matchesTypes: ['Text', 'text/plain']
            }
        };
    }
});
