// @my-react-app$clayui%2Fcore@3.87.0/lib/vertical-bar/Bar.js
"use strict";

Liferay.Loader.define("@my-react-app$clayui/core@3.87.0/lib/vertical-bar/Bar", ['module', 'exports', 'require', '@my-react-app$clayui/shared', 'my-react-app$classnames', 'my-react-app$react', '../collection', './context'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    function _typeof(obj) {
      "@babel/helpers - typeof";
      return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) {
        return typeof obj;
      } : function (obj) {
        return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
      }, _typeof(obj);
    }

    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    exports.Bar = Bar;

    var _shared = require("@my-react-app$clayui/shared");

    var _classnames = _interopRequireDefault(require("my-react-app$classnames"));

    var _react = _interopRequireWildcard(require("my-react-app$react"));

    var _collection = require("../collection");

    var _context = require("./context");

    var _excluded = ["children"];

    function _getRequireWildcardCache(nodeInterop) {
      if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
      })(nodeInterop);
    }

    function _interopRequireWildcard(obj, nodeInterop) {
      if (!nodeInterop && obj && obj.__esModule) {
        return obj;
      }if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {
        return { default: obj };
      }var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {
        return cache.get(obj);
      }var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
          var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {
            Object.defineProperty(newObj, key, desc);
          } else {
            newObj[key] = obj[key];
          }
        }
      }newObj.default = obj;if (cache) {
        cache.set(obj, newObj);
      }return newObj;
    }

    function _interopRequireDefault(obj) {
      return obj && obj.__esModule ? obj : { default: obj };
    }

    function _extends() {
      _extends = Object.assign ? Object.assign.bind() : function (target) {
        for (var i = 1; i < arguments.length; i++) {
          var source = arguments[i];for (var key in source) {
            if (Object.prototype.hasOwnProperty.call(source, key)) {
              target[key] = source[key];
            }
          }
        }return target;
      };return _extends.apply(this, arguments);
    }

    function _objectWithoutProperties(source, excluded) {
      if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {
          key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];
        }
      }return target;
    }

    function _objectWithoutPropertiesLoose(source, excluded) {
      if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {
        key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];
      }return target;
    }

    var List = function List(_ref) {
      var children = _ref.children,
          otherProps = _objectWithoutProperties(_ref, _excluded);

      return (/*#__PURE__*/_react.default.createElement("ul", _extends({}, otherProps, {
          className: "tbar-nav"
        }), children)
      );
    };

    function Bar(_ref2) {
      var children = _ref2.children,
          _ref2$displayType = _ref2.displayType,
          displayType = _ref2$displayType === void 0 ? 'dark' : _ref2$displayType,
          items = _ref2.items,
          virtualize = _ref2.virtualize;
      var parentRef = (0, _react.useRef)(null);

      var _useContext = (0, _react.useContext)(_context.VerticalBarContext),
          activation = _useContext.activation;

      var _useNavigation = (0, _shared.useNavigation)({
        activation: activation,
        containerRef: parentRef,
        orientation: 'vertical'
      }),
          navigationProps = _useNavigation.navigationProps;

      return (/*#__PURE__*/_react.default.createElement("nav", _extends({}, navigationProps, {
          className: (0, _classnames.default)('tbar tbar-stacked c-slideout-show', {
            'tbar-dark-l2': displayType === 'dark',
            'tbar-light': displayType === 'light'
          }),
          ref: parentRef
        }), /*#__PURE__*/_react.default.createElement(_collection.Collection, {
          "aria-orientation": "vertical",
          as: List,
          estimateSize: 40,
          items: items,
          parentRef: parentRef,
          role: "tablist",
          virtualize: virtualize
        }, children))
      );
    }
  }
});
