// @my-react-app$clayui%2Fform@3.87.0/lib/index.js
"use strict";

Liferay.Loader.define("@my-react-app$clayui/form@3.87.0/lib/index", ['module', 'exports', 'require', './Checkbox', './DualListBox', './Form', './Input', './Radio', './RadioGroup', './Select', './SelectBox', './SelectWithOption', './Toggle'], function (module, exports, require) {
  var define = undefined;
  var global = window;
  {
    Object.defineProperty(exports, "__esModule", {
      value: true
    });
    Object.defineProperty(exports, "ClayCheckbox", {
      enumerable: true,
      get: function get() {
        return _Checkbox.default;
      }
    });
    Object.defineProperty(exports, "ClayDualListBox", {
      enumerable: true,
      get: function get() {
        return _DualListBox.default;
      }
    });
    Object.defineProperty(exports, "ClayInput", {
      enumerable: true,
      get: function get() {
        return _Input.default;
      }
    });
    Object.defineProperty(exports, "ClayRadio", {
      enumerable: true,
      get: function get() {
        return _Radio.default;
      }
    });
    Object.defineProperty(exports, "ClayRadioGroup", {
      enumerable: true,
      get: function get() {
        return _RadioGroup.default;
      }
    });
    Object.defineProperty(exports, "ClaySelect", {
      enumerable: true,
      get: function get() {
        return _Select.default;
      }
    });
    Object.defineProperty(exports, "ClaySelectBox", {
      enumerable: true,
      get: function get() {
        return _SelectBox.default;
      }
    });
    Object.defineProperty(exports, "ClaySelectWithOption", {
      enumerable: true,
      get: function get() {
        return _SelectWithOption.default;
      }
    });
    Object.defineProperty(exports, "ClayToggle", {
      enumerable: true,
      get: function get() {
        return _Toggle.default;
      }
    });
    exports.default = void 0;

    var _Checkbox = _interopRequireDefault(require("./Checkbox"));

    var _DualListBox = _interopRequireDefault(require("./DualListBox"));

    var _Form = _interopRequireDefault(require("./Form"));

    var _Input = _interopRequireDefault(require("./Input"));

    var _Radio = _interopRequireDefault(require("./Radio"));

    var _RadioGroup = _interopRequireDefault(require("./RadioGroup"));

    var _Select = _interopRequireDefault(require("./Select"));

    var _SelectBox = _interopRequireDefault(require("./SelectBox"));

    var _SelectWithOption = _interopRequireDefault(require("./SelectWithOption"));

    var _Toggle = _interopRequireDefault(require("./Toggle"));

    function _interopRequireDefault(obj) {
      return obj && obj.__esModule ? obj : { default: obj };
    }

    /**
     * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
     * SPDX-License-Identifier: BSD-3-Clause
     */
    var _default = _Form.default;
    exports.default = _default;
  }
});
